### PR TITLE
Mostly fix metadata_only backend and extract some code out of rustc_codegen_llvm

### PR DESCRIPTION
**Pull request by bjorn3**
Removes dependency on the `ar` crate and removes the `llvm.enabled` config option in favour of setting `rust.codegen-backends` to `[]`.

### CODE REVIEW

File: src/librustc_driver/lib.rs [hunk 1]
```diff
@@ -375,10 +375,20 @@ fn get_codegen_sysroot(backend_name: &str) -> fn() -> Box<CodegenBackend> {
...
-            early_error(ErrorOutputType::default(), &err);
+                early_error(ErrorOutputType::default(), &err);
+            } else {
+                let warn = format!("no codegen-backend `{}`, \
+                               no appropriate dylib in `{}`. \
+                               Falling back to metadata_only codegen backend. \
+                               **This is suitable for dev purposes only**",
+                               backend_name, sysroot.display());
+                early_warn(ErrorOutputType::default(), &warn);
+                return rustc_codegen_utils::codegen_backend::MetadataOnlyCodegenBackend::new;
```
Comment (bjorn3) [inline src/librustc_driver/lib.rs]: I couldn't find a reliable way to pass `-Zcodegen-backend=metadata-only` to stage1+, but not the bootstrap compiler.

File: src/librustc_codegen_utils/llvm_target_features.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
+use rustc::session::Session;
```
Comment (bjorn3) [inline src/librustc_codegen_utils/llvm_target_features.rs]: Moved to `rustc_codegen_utils`, because `MetadataOnlyCodegenBackend` needs to pass them to `target_features_whitelist` to compile `core`.

File: src/librustc_codegen_utils/codegen_backend.rs [hunk 3]
```diff
@@ -81,96 +76,27 @@ pub trait CodegenBackend {
...
-        while let Some(entry_result) = archive.next_entry() {
-            let mut entry = entry_result
-                .map_err(|e| format!("metadata section read err: {:?}", e))?;
-            if entry.header().identifier() == "rust.metadata.bin" {
-                let mut buf = Vec::new();
-                io::copy(&mut entry, &mut buf).unwrap();
-                let buf: OwningRef<Vec<u8>, [u8]> = OwningRef::new(buf).into();
-                return Ok(rustc_erase_owner!(buf.map_owner_box()));
-            }
-        }
```
Comment (bjorn3) [inline src/librustc_codegen_utils/codegen_backend.rs]: It couldn't load all metadata produced by `rustc_codegen_llvm` anyway, because it didn't support dylibs.

File: src/librustc_codegen_utils/codegen_backend.rs [hunk 4]
```diff
@@ -81,96 +76,27 @@ pub trait CodegenBackend {
     ) -> Result<(), CompileIncomplete>;
 }
 
-pub struct DummyCodegenBackend;
```
Comment (bjorn3) [inline src/librustc_codegen_utils/codegen_backend.rs]: It isn't used anywere

File: src/librustc_codegen_llvm/lib.rs [hunk 5]
```diff
@@ -79,6 +79,7 @@ use rustc::ty::{self, TyCtxt};
 use rustc::util::nodemap::{FxHashSet, FxHashMap};
 use rustc_mir::monomorphize;
 use rustc_codegen_utils::codegen_backend::CodegenBackend;
+use rustc_codegen_utils::time_graph;
```
Comment (bjorn3) [inline src/librustc_codegen_llvm/lib.rs]: Moved to `rustc_codegen_utils`, because it is useful for other future backends too.

File: src/bootstrap/compile.rs [hunk 6]
```diff
@@ -640,7 +642,9 @@ impl Step for CodegenBackend {
             .arg(builder.src.join("src/librustc_codegen_llvm/Cargo.toml"));
         rustc_cargo_env(builder, &mut cargo);
 
-        features += &build_codegen_backend(&builder, &mut cargo, &compiler, target, backend);
+        if !backend.is_empty() {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/compile.rs]: This is probably not the check you wanted -- we shouldn't be creating `CodegenBackend` with an empty string. I think we can probably just remove this if.

File: src/librustc_codegen_llvm/base.rs [hunk 7]
```diff
@@ -715,7 +714,7 @@ pub fn codegen_crate<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                              rx: mpsc::Receiver<Box<Any + Send>>)
                              -> OngoingCodegen {
 
-    check_for_rustc_errors_attr(tcx);
+    ::rustc_codegen_utils::check_for_rustc_errors_attr(tcx);
```
Comment (Mark-Simulacrum) [inline src/librustc_codegen_llvm/base.rs]: What's the advantage to not importing this?

**Review by Mark-Simulacrum — COMMENTED** [review]
Rustbuild changes look acceptable; could you add a comment to config.toml.example about disabling LLVM via the codegen-backends array?

File: src/librustc_codegen_llvm/base.rs (hunk 7 above)
Comment (bjorn3) [reply src/librustc_codegen_llvm/base.rs]: Nothing

File: src/bootstrap/compile.rs (hunk 6 above)
Comment (bjorn3) [reply src/bootstrap/compile.rs]: What exactly do you mean?

Comment (Mark-Simulacrum) [reply src/bootstrap/compile.rs]: `backend` here isn't an array, but a `Interned<String>` -- this check should be removed.

**Comment by bjorn3** [comment]
Why is it failing?

**Comment by bjorn3** [comment]
Still not fixed :(

**Comment by bjorn3** [comment]
Fixed :)

**Comment by bjorn3** [comment]
Squashed the commits to fix the test

File: src/librustc_codegen_utils/codegen_backend.rs [hunk 8]
```diff
@@ -225,7 +161,7 @@ impl CodegenBackend for MetadataOnlyCodegenBackend {
                 collector::MonoItemCollectionMode::Eager
             ).0.iter()
         );
-        ::rustc::middle::dependency_format::calculate(tcx);
+        //::rustc::middle::dependency_format::calculate(tcx);
```
Comment (oli-obk) [inline src/librustc_codegen_utils/codegen_backend.rs]: commented out code?

Comment (bjorn3) [reply src/librustc_codegen_utils/codegen_backend.rs]: Crashes atm.

Comment (bjorn3) [reply src/librustc_codegen_utils/codegen_backend.rs]: Added a FIXME

File: config.toml.example [hunk 9]
```diff
@@ -334,8 +330,8 @@
 # This is an array of the codegen backends that will be compiled for the rustc
 # that's being compiled. The default is to only build the LLVM codegen backend,
 # but you can also optionally enable the "emscripten" backend for asm.js or
-# make this an empty array (but that probably won't get too far in the
-# bootstrap)
+# make this an empty array (that will disable LLVM, but bootstrap will fail at
+# the time the final rustc binary is built for stage 1)
```
Comment (bjorn3) [inline config.toml.example]: Is this the right stage? I always get confused about it.

File: src/librustc_driver/lib.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_driver/lib.rs]: Isn't it `RUSTFLAGS_STAGE_NOT_0`?

File: src/librustc_codegen_utils/codegen_backend.rs (hunk 4 above)
Comment (eddyb) [reply src/librustc_codegen_utils/codegen_backend.rs]: Not even in `librustc_driver/test.rs`?

File: src/librustc_codegen_utils/llvm_target_features.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_codegen_utils/llvm_target_features.rs]: I don't think this is a good direction - LLVM shouldn't be treated that special. Is there a way to make `core` not contain a usage `core::arch` if the target features are missing?

File: config.toml.example (hunk 9 above)
Comment (eddyb) [reply config.toml.example]: I don't think this needs to be that explicit, keeping the original "but that probably won't get too far in the bootstrap" and just adding "which will disable LLVM" before it seems enough.

File: src/librustc_codegen_llvm/common.rs [hunk 10]
```diff
@@ -448,3 +448,101 @@ pub fn ty_fn_sig<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>,
         _ => bug!("unexpected type {:?} to ty_fn_sig", ty)
     }
 }
+
+pub fn size_and_align_of_dst<'a, 'tcx>(bx: &Builder<'a, 'tcx>, t: Ty<'tcx>, info: ValueRef)
```
Comment (eddyb) [inline src/librustc_codegen_llvm/common.rs]: Why move this? I'd rather see `base` and `common` go away just as much as `glue`, so moving something here doesn't seem that useful. IMO this should be a method on `PlaceRef`, that's just `size_and_align_of`.

File: src/librustc_codegen_utils/codegen_backend.rs [hunk 11]
```diff
@@ -280,9 +217,8 @@ impl CodegenBackend for MetadataOnlyCodegenBackend {
             } else {
                 &ongoing_codegen.metadata.raw_data
             };
-            let mut builder = Builder::new(File::create(&output_name).unwrap());
-            let header = Header::new("rust.metadata.bin".to_string(), metadata.len() as u64);
-            builder.append(&header, Cursor::new(metadata)).unwrap();
+            let mut file = File::create(&output_name).unwrap();
+            file.write_all(metadata).unwrap();
```
Comment (eddyb) [inline src/librustc_codegen_utils/codegen_backend.rs]: So all library types are effectively the same as `--emit=metadata`?

**Comment by eddyb** [comment]
In general I dislike `rustc_codegen_utils` and would prefer if its parts were moved into `rustc_driver` (i.e. the `CodegenBackend` trait specifically), `rustc_mir` (if at all applicable), or somewhere in `rustc`.

File: src/librustc_codegen_utils/codegen_backend.rs (hunk 11 above)
Comment (bjorn3) [reply src/librustc_codegen_utils/codegen_backend.rs]: Yes

File: src/Cargo.lock [hunk 12]
```diff
@@ -2110,7 +2110,6 @@ dependencies = [
 name = "rustc_codegen_utils"
 version = "0.0.0"
 dependencies = [
- "ar 0.3.1 (registry+https://github.com/rust-lang/crates.io-index)",
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/Cargo.lock]: No

File: src/librustc_codegen_utils/codegen_backend.rs [hunk 13]
```diff
@@ -22,15 +22,13 @@
...
 use std::fs::File;
 use std::path::Path;
-use std::sync::mpsc;
+use std::sync::{mpsc, Arc};
 
 use rustc_data_structures::owning_ref::OwningRef;
 use rustc_data_structures::sync::Lrc;
-use ar::{Archive, Builder, Header};
 use flate2::Compression;
 use flate2::write::DeflateEncoder;
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/librustc_codegen_utils/codegen_backend.rs]: Compressing the metadata, when rustc_metadata expects it.

File: src/Cargo.lock (hunk 12 above)
[the remark this answers is not in the document]
Comment (eddyb) [reply src/Cargo.lock]: I was asking because there's a "checksum" line that would also be removed if this was the last use of `ar`.
I just checked and `rustc_driver` also depends on `ar`.

[the remark this answers is not in the document]
Comment (bjorn3) [reply src/Cargo.lock]: what??? how did tidy pass then? i removed ar as whitelisted dep

File: src/librustc_codegen_utils/codegen_backend.rs (hunk 4 above)
Comment (bjorn3) [reply src/librustc_codegen_utils/codegen_backend.rs]: No

File: src/Cargo.lock (hunk 12 above)
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/Cargo.lock]: Removed unused dep

**Comment by bjorn3** [comment]
> In general I dislike rustc_codegen_utils and would prefer if its parts were moved into rustc_driver (i.e. the CodegenBackend trait specifically), rustc_mir (if at all applicable), or somewhere in rustc.

Is it okay to do that in a next pr?

**Comment by eddyb** [comment]
> Is it okay to do that in a next pr?

Alright, but can you move `time_graph.rs` to `librustc/util`? Also, I'm not sure where to put the target feature whitelist. Or if we should have one at all for non-LLVM backends. cc @gnzlbg

**Comment by bjorn3** [comment]
> Alright, but can you move time_graph.rs to librustc/util?

Sure

> Also, I'm not sure where to put the target feature whitelist. Or if we should have one at all for non-LLVM backends.

I want to cfg the core module needing them out if not compiled using LLVM.

File: src/librustc_codegen_utils/llvm_target_features.rs (hunk 2 above)
Comment (gnzlbg) [reply src/librustc_codegen_utils/llvm_target_features.rs]: @eddyb what do you mean here with "if the target features are missing"? Because of run-time feature detection, `core::arch` always need to be available.

> I don't think this is a good direction - LLVM shouldn't be treated that special

Once we add cretonne as a backend we'll need to map Rust's feature names to backend names (once for LLVM, once for Cretonne), and all backends won't support all targets. So I think that at least the white-lists are going to need to exist on a per backend basis.

File: src/test/run-make-fulldeps/hotplug_codegen_backend/Makefile [hunk 14]
```diff
@@ -4,6 +4,6 @@ all:
 	/bin/echo || exit 0 # This test requires /bin/echo to exist
 	$(RUSTC) the_backend.rs --crate-name the_backend --crate-type dylib \
 		-o $(TMPDIR)/the_backend.dylib
-	$(RUSTC) some_crate.rs --crate-name some_crate --crate-type bin -o $(TMPDIR)/some_crate \
+	$(RUSTC) some_crate.rs --crate-name some_crate --crate-type lib -o $(TMPDIR)/some_crate \
 		-Z codegen-backend=$(TMPDIR)/the_backend.dylib -Z unstable-options
```
Comment (gnzlbg) [inline src/test/run-make-fulldeps/hotplug_codegen_backend/Makefile]: Shouldn't we have two tests here, one for `lib` and one for `bin` ?

File: src/librustc_codegen_utils/codegen_backend.rs [hunk 15]
```diff
@@ -161,7 +161,8 @@ impl CodegenBackend for MetadataOnlyCodegenBackend {
                 collector::MonoItemCollectionMode::Eager
             ).0.iter()
         );
-        //::rustc::middle::dependency_format::calculate(tcx);
+        // FIXME: Fix this
+        // ::rustc::middle::dependency_format::calculate(tcx);
```
Comment (gnzlbg) [inline src/librustc_codegen_utils/codegen_backend.rs]: Could the fixme explain what needs to be fixed? Why is that necessary and why doesn't it work?

File: config.toml.example [hunk 16]
```diff
@@ -330,8 +330,8 @@
 # This is an array of the codegen backends that will be compiled for the rustc
 # that's being compiled. The default is to only build the LLVM codegen backend,
 # but you can also optionally enable the "emscripten" backend for asm.js or
-# make this an empty array (that will disable LLVM, but bootstrap will fail at
-# the time the final rustc binary is built for stage 1)
+# make this an empty array (which will disable LLVM, but that probably won't
+# get too far in the bootstrap
```
Comment (gnzlbg) [inline config.toml.example]: missing closing parent `)`.

**Comment by gnzlbg** [comment]
> Also, I'm not sure where to put the target feature whitelist. Or if we should have one at all for non-LLVM backends. cc @gnzlbg

I think we need to map Rust feature names to the backend names for each backend, so at the end we should have one feature list for each backend, even if for some backends it is empty. 

As more backends supporting many features are added (probably Cretonne will be the first one), we will probably want to have the code to generate the feature lists in one place, mapping from rust feature names to each backend feature name on the same line. 

For now, only LLVM does something meaningful with features, and the lists of other backends like `emscripten` should probably be empty, so I don't mind about having one list per file. It might not be the best direction, but we probably can't know what the best direction is until we need to deal with Cretonne anyways :/

File: src/test/run-make-fulldeps/hotplug_codegen_backend/Makefile (hunk 14 above)
Comment (bjorn3) [reply src/test/run-make-fulldeps/hotplug_codegen_backend/Makefile]: Bin doesn't work without adding several lang items. `some_crate.rs` is `#![no_core]`, because `libcore` and `libstd` use `ar` based rlibs, while this pr replaces them with plain metadata for `MetadataOnlyCodegenBackend`, because it didn't support dylibs anyway.

File: config.toml.example (hunk 16 above)
Comment (bjorn3) [reply config.toml.example]: Fixed

File: src/librustc_codegen_utils/codegen_backend.rs (hunk 15 above)
Comment (bjorn3) [reply src/librustc_codegen_utils/codegen_backend.rs]: It ices

**Comment by bjorn3** [comment]
Looks like I broke stdsimd :(

**Comment by pietroalbini** [comment]
Restarted the Travis CI build, the failure was caused by the crates.io outage.

**Comment by bjorn3** [comment]
@TimNN empty log

**Comment by eddyb** [comment]
cc @alexcrichton If you're back, you might be interested in this PR, too.

File: src/librustc/session/config.rs [hunk 17]
```diff
@@ -1412,6 +1412,18 @@ pub fn default_configuration(sess: &Session) -> ast::CrateConfig {
     if sess.opts.crate_types.contains(&CrateTypeProcMacro) {
         ret.insert((Symbol::intern("proc_macro"), None));
     }
+    /*if nightly_options::is_nightly_build() {
```
Comment (alexcrichton) [inline src/librustc/session/config.rs]: Stray commented block?

**Comment by alexcrichton** [comment]
Indeed, thanks @eddyb  for the cc and @bjorn3 for the PR!

File: src/librustc/session/config.rs (hunk 17 above)
Comment (bjorn3) [reply src/librustc/session/config.rs]: I know this PR is currently broken. Will fix it soon.

**Comment by bjorn3** [comment]
I can't get partial bootstrap without LLVM working without breaking stdsimd, so I reverted those changes. I will submit a new PR when I get it working.

This is ready for review.

**Comment by bjorn3** [comment]
Rebased, because Cargo.lock had conflicts.

**Comment by gnzlbg** [comment]
Why was `stdsimd` breaking?

**Comment by bjorn3** [comment]
Because my `#[cfg]` to disable it when using `metadata_only` backend somehow also disabled with when using LLVM.

**Comment by stokhos** [comment]
Ping from triage, @eddyb @alexcrichton  this PR needs your review.

**Comment by bjorn3** [comment]
Rebased
